tensorflow/compiler/xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -219,49 +203,222 @@ static StatusOr<absl::optional<se::blas::AlgorithmType>> DoUncachedGemmAutotune(
+  int64_t rng_state = 0;
+  auto get_initialized_buffer =
+      [&](const HloInstruction* op) -> StatusOr<se::DeviceMemoryBase> {
+    TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase buffer,
+                        input_output_allocator.AllocateBytes(
+                            ShapeUtil::ByteSizeOf(op->shape())));
+    if (init_cublas_data) {
+      InitializeBuffer(stream, op->shape().element_type(), &rng_state, buffer);
+    }
+    return buffer;
+  };
 
-  tensorflow::mutex_lock cache_lock(autotune_cache_mu);
-  auto it = autotune_cache.find(key);
-  int64_t autotuning_requests = cache_hits + cache_misses;
-  if (autotuning_requests && autotuning_requests % 10 == 0) {
-    VLOG(2) << "Autotuning cache hits/(hits + misses): " << cache_hits << "/"
-            << autotuning_requests;
-  }
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase lhs_buffer,
+                      get_initialized_buffer(instr->operand(0)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase rhs_buffer,
+                      get_initialized_buffer(instr->operand(1)));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase output_buffer,
+                      get_initialized_buffer(instr));
+  TF_ASSIGN_OR_RETURN(se::DeviceMemoryBase reference_result_buffer,
+                      get_initialized_buffer(instr));
+
+  int64_t batch_size = gemm_config.batch_size();
+  GpuGemmConfig config = GetGpuGemmConfig(instr);
+  const Shape& output_shape = config.output_shape;
+  absl::flat_hash_set<PrimitiveType> enabled_types = {F16, F32, F64, C64, C128};
+
+  if (config.use_cublaslt && stream->parent()->SupportsBlasPlans() &&
+      enabled_types.contains(output_shape.element_type())) {
+    se::blas::MatrixDescriptor lhs_matrix, rhs_matrix, output_matrix;
+    std::tie(lhs_matrix, rhs_matrix, output_matrix) =
+        PopulateInputOutputMatrices(config, lhs_buffer, rhs_buffer,
+                                    output_buffer);
+
+    if (output_matrix.transpose != se::blas::Transpose::kNoTranspose) {
+      return InternalError("GEMM output matrix must not be transposed.");
+    }
+    tensorflow::DataType dtype;
+    se::blas::DataType blas_dtype;
+    switch (output_shape.element_type()) {
+      case F16:
+        dtype = tensorflow::DataTypeToEnum<Eigen::half>::value;
+        blas_dtype = se::blas::ToDataType<Eigen::half>::value;
+        break;
+      case F32:
+        dtype = tensorflow::DataTypeToEnum<float>::value;
+        blas_dtype = se::blas::ToDataType<float>::value;
+        break;
+      case F64:
+        dtype = tensorflow::DataTypeToEnum<double>::value;
+        blas_dtype = se::blas::ToDataType<double>::value;
+        break;
+      case C64:
+        dtype = tensorflow::DataTypeToEnum<complex64>::value;
+        blas_dtype = se::blas::ToDataType<complex64>::value;
+        break;
+      case C128:
+        dtype = tensorflow::DataTypeToEnum<complex128>::value;
+        blas_dtype = se::blas::ToDataType<complex128>::value;
```

Comment:
Needing blas_dtype at this abstraction level seems wrong, why can't we use tensorflow::DataType and convert it inside blas::DataType when needed inside SE? Then we can also factor this out into a function PrimitiveType->tensorflow::DataType (and probably reuse another function since it seems we have quite a lot of those already)